File: golang/internal/metrics/metrics.go
// Package metrics defines a small, OpenTelemetry-free interface for the
// three operational metrics this agent records (events received, events
// dropped, and render latency). It exists so internal/host (and, through
// it, internal/pipeline and internal/aggregator) never has to import any
// go.opentelemetry.io/otel* package -- following the same "define the
// interface in internal/*, let the head that needs a real implementation
// supply it" pattern already used by identity.Provider, toast.Renderer, and
// natsauth.Provider (see e.g. internal/identity's package doc). Only
// cmd/notify-agent-windows constructs a real OpenTelemetry-backed
// AgentMetrics; cmd/notify-agent-console never touches this package's
// NullAgentMetrics is used unconditionally, and internal/host defaults to it
// whenever no metrics implementation is supplied.
package metrics

// AgentMetrics records the three critical operational metrics for this
// agent. Implementations (e.g. an OpenTelemetry-backed one, supplied by the
// Windows head) must never panic -- a metrics-recording failure must never
// be allowed to interrupt event processing, dedup, aggregation, or
// acknowledgement publishing. Every caller in this repository additionally
// wraps every call to these methods in its own defer/recover as an explicit
// extra safety net (see internal/host's safeRecord), on top of whatever
// panic-safety an implementation provides internally -- this is a
// deliberate, unusual, user-mandated requirement: metrics-recording code
// must never be able to crash the application, even via an unrecovered
// panic somehow escaping the implementation itself.
type AgentMetrics interface {
	// RecordEventReceived fires once per valid, first-seen event accepted
	// into the pipeline -- the same point the observed_by_agent
	// acknowledgement is published.
	RecordEventReceived()

	// RecordEventDropped fires once per dropped event. reason is
	// "queue_full" or "bucket_overflow", matching the existing
	// DroppedQueueFull/DroppedBucketOverflow counters this call sits
	// alongside.
	RecordEventDropped(reason string)

	// RecordRenderDuration fires once per event represented in a
	// successfully rendered toast (once per event in a batch, not once per
	// toast), with that event's own end-to-end agent processing latency in
	// seconds (toastSubmittedAt - that event's AgentReceivedAt).
	RecordRenderDuration(seconds float64)
}

// NullAgentMetrics is a no-op AgentMetrics. Used by the console head (which
// never configures metrics), by internal/host/pipeline/aggregator whenever
// no AgentMetrics is supplied, and as the Windows head's own fallback if
// OpenTelemetry setup fails -- so metrics-recording call sites never need a
// nil check of their own.
type NullAgentMetrics struct{}

// RecordEventReceived implements AgentMetrics.
func (NullAgentMetrics) RecordEventReceived() {}

// RecordEventDropped implements AgentMetrics.
func (NullAgentMetrics) RecordEventDropped(reason string) {}

// RecordRenderDuration implements AgentMetrics.
func (NullAgentMetrics) RecordRenderDuration(seconds float64) {}

File: rust/notify-agent-core/src/metrics.rs
//! Metrics-recording trait for the agent's three critical operational
//! counters/histogram. Mirrors the established pattern for pluggable,
//! host-supplied behavior in this crate (`IdentityProvider`, `ToastRenderer`,
//! `NatsAuthProvider`): a small trait lives here in Core, a trivial no-op
//! default lives here too, and any real implementation (e.g. an
//! OpenTelemetry-backed one) is supplied by whichever head needs it — see
//! `notify-agent-windows/src/otel_metrics.rs`. This deliberately keeps the
//! `opentelemetry`/`opentelemetry-otlp` crates out of this crate's (and the
//! console head's) dependency tree.
//!
//! # Crash-safety contract
//!
//! Every method here returns `()`, not `Result` — recording a metric is
//! never allowed to be a fallible operation from the caller's point of view.
//! Implementations MUST NOT panic: no `.unwrap()`/`.expect()` on anything
//! that can fail from external conditions (a bad endpoint, a network issue,
//! an SDK error), no arithmetic overflow, no out-of-bounds indexing. Any
//! internal failure must be swallowed (optionally logged via
//! `tracing::warn!`) rather than propagated, so a metrics-recording failure
//! can never interrupt event processing, dedup, aggregation, or
//! acknowledgement publishing (the actual business logic this trait's call
//! sites sit next to in `pipeline.rs`/`aggregator.rs`).
pub trait AgentMetrics: Send + Sync {
    /// One valid, first-seen event was accepted into the pipeline (fires at
    /// the same point the observed_by_agent acknowledgement does).
    fn record_event_received(&self);

    /// One event was dropped. `reason` is `"queue_full"` or
    /// `"bucket_overflow"` (matching the existing dropped_queue_full /
    /// dropped_bucket_overflow counters this call sits alongside).
    fn record_event_dropped(&self, reason: &str);

    /// One event's end-to-end agent processing latency, in seconds
    /// (toast_submitted_at - agent_received_at for that specific event).
    /// Fires once per event represented in a successfully rendered toast —
    /// once per event in a batch, not once per toast.
    fn record_render_duration(&self, seconds: f64);
}

/// No-op default. Used by the console head (which never configures
/// metrics) and as the Windows head's own fallback if OpenTelemetry setup
/// fails, so metrics-recording call sites in core never need an `Option`
/// check — there is always some `Arc<dyn AgentMetrics>` to call.
pub struct NullAgentMetrics;

impl AgentMetrics for NullAgentMetrics {
    fn record_event_received(&self) {}
    fn record_event_dropped(&self, _reason: &str) {}
    fn record_render_duration(&self, _seconds: f64) {}
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn null_agent_metrics_never_panics() {
        let m = NullAgentMetrics;
        m.record_event_received();
        m.record_event_dropped("queue_full");
        m.record_event_dropped("bucket_overflow");
        m.record_render_duration(1.5);
    }
}
